Add page render functionality (#24)
Artemis21/summer-code-jam-2020: jolly-jellyfish/README.md

# Jolly Jellyfish
![Jolly Jellyfish logo](team_logo.svg)

In today's fast moving society, you may want to have your own web property along the information superhighway... 
However, with the latest shutdown of Yahoo!'s Geocities, where to? Don't worry! Halfway™️ has got your back.
Initially named as a joke on Medium, we've expanded into a profitable enterprise.
However, our angel investors did not quite believe us and left us for the next Google...
Now that we're bankrupt, we have a cultural obligation to opensource this.

## Idea

An old-internet style site-creation tool that comes with the kitchen sink.

Simply put, users can choose from a variety of templates and customize it any-which-way.
After this, they can take their site to the forum to get ~~likes~~ feedback.
The most liked sites appear on the front page, along with the most liked users and most recent sites.

## Initial Setup

> ### wkhtmltoimage

Reviewer: Did you forget to update this heading?

Author: It appears so!

> ### wkhtmltoimage

Author: ```suggestion
### Selenium
```

In order to generate page thumbnails (using `selenium`), a (platform specific) browser driver binary is required (we use the _Chrome/Firefox_ drivers).
Download the image for your Chrome/Firefox version (most likely '_Current stable release_') from here: [Chrome/Chromium][1]; [Firefox][2] 
and set the system environment variable **SELENIUM_DRIVER** to the binary's path. (This is obviously assuming you have Chrome/Firefox installed).

E.g. On Windows (admin) with the chromedriver.exe file, this is done with:
```batch
...\> setx SELENIUM_DRIVER "C:\Path\to\chromedriver.exe" /m
```
(Use `SETENV` command on UNIX)

#### Notes on 'real' distribution
In a production environment, this process would obviously not be needed as the server would be running a known OS.
Therefore, only one executable would need to be download to the src directory and then hard linked within the code; rather than having to use `os.environ`.

Please also note that, as described within `src/page_maker/templatetags/render_thumbnail.py`, 
asynchronous tasks (using `Celery` and `Redis`) are not easily setup on Windows - 

> hence this feature can cause loading times of ~5-10 seconds after template and webpage creation forms are initially submitted.

Reviewer: Could we drop a `FIXME: find if we can use docker for celery / redis` before this (I think PyCharm will catch that) so that we can see if there's a way before finishing the project?

Author: Baru did mention potentially setting one up in the Discord.

> 

Author: ```suggestion
FIXME: find if we can use docker for celery / redis

```

> ### pipenv

Reviewer: ```suggestion
### Using pipenv
```

Then we can add a `using <xyz>` if we want to use something else.

> ```shell script

Reviewer: Should we use `sh` for syntax highlighting on GitHub?

Author: Yeah, I think this was just pycharm autocompleting it for me.

> ```shell script

Author: ```suggestion
```sh
```

> $ pipenv install

Author: ```suggestion
$ cd jolly-jellyfish
$ pipenv install
```

$ pipenv shell

> (venv) $ python manage.py makemigrations  # TODO: this might be optional, depends on pipenv script section.

Reviewer: My intent by "this might be optional" is that the migrations might already be made for the final application, meaning we don't have to run `makemigrations`.

Author: Oh ok - I get you.
```suggestion
(venv) $ python manage.py makemigrations  # TODO: this might be optional, depends on migrations already run.
```

> (venv) $ python manage.py migrate

Reviewer: Don't forget to `cd src` first! (or specify your directory before the shell session)

```

## Usage

> ```shell script

Reviewer: Should we use `sh` for syntax highlighting on GitHub?

> ```shell script

Author: ```suggestion
```sh
(venv) $ cd src
```

> (venv) $ python manage.py runserver

Reviewer: Don't forget to `cd src` or specify the directory you are in.

```

[1]: https://sites.google.com/a/chromium.org/chromedriver/home

> [2]: https://github.com/mozilla/geckodriver/releases

Reviewer: Could we move these footnotes up? (Don't really know much about markdown footnotes :p)... It's annoying you have to scroll so far down to get context in plaintext.

Author: In the actual README these links are obviously clickable in their original location so it's not too much of an issue where these footnotes are.

Please also note that, as described within `src/page_maker/templatetags/render_thumbnail.py`, 
asynchronous tasks (using `Celery` and `Redis`) are not easily setup on Windows - 
hence this feature can cause loading times of ~5-10 seconds after template and webpage creation forms are initially submitted.
FIXME: find if we can use docker for celery / redis

### Using pipenv
```sh
$ cd src

> $ cd jolly-jellyfish

Reviewer: Whoops, ordered incorrectly: either remove `cd jolly-jellyfish` or reverse the ordering.
